sql/stats: TestRandomQuantileRoundTrip failed

sql/stats.TestRandomQuantileRoundTrip [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16815025?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16815025?buildTab=artifacts#/) on master @ [942d2f819a7eb89a5166bef95274b9d8f438dacb](https://github.com/cockroachdb/cockroach/commits/942d2f819a7eb89a5166bef95274b9d8f438dacb):


```
=== RUN   TestRandomQuantileRoundTrip
--- FAIL: TestRandomQuantileRoundTrip (0.01s)
=== RUN   TestRandomQuantileRoundTrip/timestamptz/2
    --- FAIL: TestRandomQuantileRoundTrip/timestamptz/2 (0.00s)
panic: timestamp "-4713-06-15T10:10:10Z" exceeds supported timestamp bounds [recovered]
	panic: timestamp "-4713-06-15T10:10:10Z" exceeds supported timestamp bounds

goroutine 7769 [running]:
testing.tRunner.func1.2({0x4a2b940, 0x400458c940})
	GOROOT/src/testing/testing.go:1631 +0x1c4
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x33c
panic({0x4a2b940?, 0x400458c940?})
	GOROOT/src/runtime/panic.go:770 +0x124
github.com/cockroachdb/cockroach/pkg/sql/stats.randBounds(0x95f4520, 0x400a0088a0, 0xb4)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:221 +0x554
github.com/cockroachdb/cockroach/pkg/sql/stats.randHist({0x64bfd68, 0x9adada0}, 0x95f4520, 0x400a0088a0)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:76 +0x64
github.com/cockroachdb/cockroach/pkg/sql/stats.TestRandomQuantileRoundTrip.func1(0x400aa34000)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:50 +0x54
testing.tRunner(0x400aa34000, 0x400961ed50)
	GOROOT/src/testing/testing.go:1689 +0xec
created by testing.(*T).Run in goroutine 7731
	GOROOT/src/testing/testing.go:1742 +0x318
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomQuantileRoundTrip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-42019

sql/stats.TestRandomQuantileRoundTrip [failed](https://tanzanite.cluster.engflow.com/invocations/default/ea9d0068-3938-40fe-95c1-1b15bfa05892?testReportRun=1&testReportShard=6&testReportAttempt=1#targets-Ly9wa2cvc3FsL3N0YXRzOnN0YXRzX3Rlc3Q=) on master @ [285460a47358406e65df9ff3deb5a0764f58a2bf](https://github.com/cockroachdb/cockroach/commits/285460a47358406e65df9ff3deb5a0764f58a2bf):

Fatal error:

```
panic: timestamp "-4713-02-02T15:10:19Z" exceeds supported timestamp bounds [recovered]
	panic: timestamp "-4713-02-02T15:10:19Z" exceeds supported timestamp bounds

```
Stack: 

```
goroutine 6045 [running]:
testing.tRunner.func1.2({0x8240340, 0xc00280a7a0})
	GOROOT/src/testing/testing.go:1631 +0x3f7
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x6b6
panic({0x8240340?, 0xc00280a7a0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/stats.randBounds(0xeeef6a0, 0xc006de54a0, 0x7a)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:221 +0x99d
github.com/cockroachdb/cockroach/pkg/sql/stats.randHist({0xb75a3c8, 0x1048c880}, 0xeeef6a0, 0xc006de54a0)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:76 +0x96
github.com/cockroachdb/cockroach/pkg/sql/stats.TestRandomQuantileRoundTrip.func1(0xc007351040)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:50 +0x9d
testing.tRunner(0xc007351040, 0xc007ac8c60)
	GOROOT/src/testing/testing.go:1689 +0x21f
created by testing.(*T).Run in goroutine 6005
	GOROOT/src/testing/testing.go:1742 +0x826
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRandomQuantileRoundTrip
--- FAIL: TestRandomQuantileRoundTrip (0.03s)
=== RUN   TestRandomQuantileRoundTrip/timestamptz/4
    --- FAIL: TestRandomQuantileRoundTrip/timestamptz/4 (0.00s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=6</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomQuantileRoundTrip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

sql/stats.TestRandomQuantileRoundTrip [failed](https://tanzanite.cluster.engflow.com/invocations/default/c88f84f8-dcf3-4193-87ee-6129ef763fa5?testReportRun=1&testReportShard=6&testReportAttempt=1#targets-Ly9wa2cvc3FsL3N0YXRzOnN0YXRzX3Rlc3Q=) on master @ [d26f61dd5047c2590ac504adda33caf246386c0e](https://github.com/cockroachdb/cockroach/commits/d26f61dd5047c2590ac504adda33caf246386c0e):

Fatal error:

```
panic: timestamp "-4713-01-25T11:16:48Z" exceeds supported timestamp bounds [recovered]
	panic: timestamp "-4713-01-25T11:16:48Z" exceeds supported timestamp bounds

```
Stack: 

```
goroutine 5645 [running]:
testing.tRunner.func1.2({0x824ff40, 0xc00652d780})
	GOROOT/src/testing/testing.go:1631 +0x3f7
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x6b6
panic({0x824ff40?, 0xc00652d780?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/stats.randBounds(0xef0a800, 0xc004a3cf30, 0x38)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:221 +0x99d
github.com/cockroachdb/cockroach/pkg/sql/stats.randHist({0xb76dad8, 0x104a7ac0}, 0xef0a800, 0xc004a3cf30)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:76 +0x96
github.com/cockroachdb/cockroach/pkg/sql/stats.TestRandomQuantileRoundTrip.func1(0xc0044bfba0)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:50 +0x9d
testing.tRunner(0xc0044bfba0, 0xc0065f6450)
	GOROOT/src/testing/testing.go:1689 +0x21f
created by testing.(*T).Run in goroutine 5593
	GOROOT/src/testing/testing.go:1742 +0x826
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRandomQuantileRoundTrip
--- FAIL: TestRandomQuantileRoundTrip (0.03s)
=== RUN   TestRandomQuantileRoundTrip/timestamptz/0
    --- FAIL: TestRandomQuantileRoundTrip/timestamptz/0 (0.00s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=6</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomQuantileRoundTrip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

sql/stats.TestRandomQuantileRoundTrip [failed](https://tanzanite.cluster.engflow.com/invocations/default/4f634b6f-92bd-438a-9d06-0436686c3506?testReportRun=5&testReportShard=6&testReportAttempt=1#targets-Ly9wa2cvc3FsL3N0YXRzOnN0YXRzX3Rlc3Q=) on master @ [b6c13686495bbe9ad476b28033461ef7628e18a8](https://github.com/cockroachdb/cockroach/commits/b6c13686495bbe9ad476b28033461ef7628e18a8):

Fatal error:

```
panic: timestamp "-4713-04-20T04:14:22Z" exceeds supported timestamp bounds [recovered]
	panic: timestamp "-4713-04-20T04:14:22Z" exceeds supported timestamp bounds

```
Stack: 

```
goroutine 6980 [running]:
testing.tRunner.func1.2({0x5138c00, 0xc0044aafa0})
	GOROOT/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x377
panic({0x5138c00?, 0xc0044aafa0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/stats.randBounds(0xa0b4d40, 0xc0062f0ae0, 0x87)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:221 +0x5ba
github.com/cockroachdb/cockroach/pkg/sql/stats.randHist({0x6bfcdb8, 0xa715d00}, 0xa0b4d40, 0xc0062f0ae0)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:76 +0x85
github.com/cockroachdb/cockroach/pkg/sql/stats.TestRandomQuantileRoundTrip.func1(0xc001a649c0)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:50 +0x5e
testing.tRunner(0xc001a649c0, 0xc006350060)
	GOROOT/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 6944
	GOROOT/src/testing/testing.go:1742 +0x390
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRandomQuantileRoundTrip
--- FAIL: TestRandomQuantileRoundTrip (0.01s)
=== RUN   TestRandomQuantileRoundTrip/timestamptz/0
    --- FAIL: TestRandomQuantileRoundTrip/timestamptz/0 (0.00s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=5</code>
 - <code>shard=6</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomQuantileRoundTrip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

sql/stats.TestRandomQuantileRoundTrip [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17261487?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17261487?buildTab=artifacts#/) on master @ [188d9fe7211ca231b8359bef085b1b1b1301b713](https://github.com/cockroachdb/cockroach/commits/188d9fe7211ca231b8359bef085b1b1b1301b713):


```
=== RUN   TestRandomQuantileRoundTrip
--- FAIL: TestRandomQuantileRoundTrip (0.00s)
=== RUN   TestRandomQuantileRoundTrip/timestamptz/3
    --- FAIL: TestRandomQuantileRoundTrip/timestamptz/3 (0.00s)
panic: timestamp "-4714-11-29T06:29:18Z" exceeds supported timestamp bounds [recovered]
	panic: timestamp "-4714-11-29T06:29:18Z" exceeds supported timestamp bounds

goroutine 7706 [running]:
testing.tRunner.func1.2({0x4ae7a20, 0x4008c2caa0})
	GOROOT/src/testing/testing.go:1631 +0x1c4
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x33c
panic({0x4ae7a20?, 0x4008c2caa0?})
	GOROOT/src/runtime/panic.go:770 +0x124
github.com/cockroachdb/cockroach/pkg/sql/stats.randBounds(0x97966c0, 0x4001573950, 0x95)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:216 +0x554
github.com/cockroachdb/cockroach/pkg/sql/stats.randHist({0x65b67f8, 0x9c7fb20}, 0x97966c0, 0x4001573950)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:71 +0x64
github.com/cockroachdb/cockroach/pkg/sql/stats.TestRandomQuantileRoundTrip.func1(0x40059b6d00)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile_test.go:45 +0x54
testing.tRunner(0x40059b6d00, 0x4005eba5a0)
	GOROOT/src/testing/testing.go:1689 +0xec
created by testing.(*T).Run in goroutine 7667
	GOROOT/src/testing/testing.go:1742 +0x318
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomQuantileRoundTrip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

sql/stats.TestRandomQuantileRoundTrip [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17628150?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17628150?buildTab=artifacts#/) on master @ [2c3fc8e36c04f987081a55806b45b3faca0a9dc7](https://github.com/cockroachdb/cockroach/commits/2c3fc8e36c04f987081a55806b45b3faca0a9dc7):


```
=== RUN   TestRandomQuantileRoundTrip
--- FAIL: TestRandomQuantileRoundTrip (0.00s)
=== RUN   TestRandomQuantileRoundTrip/timestamptz/3
    --- FAIL: TestRandomQuantileRoundTrip/timestamptz/3 (0.00s)
panic: timestamp "-4713-04-16T08:37:57Z" exceeds supported timestamp bounds [recovered]
	panic: timestamp "-4713-04-16T08:37:57Z" exceeds supported timestamp bounds

goroutine 7118 [running]:
testing.tRunner.func1.2({0x4a573a0, 0x4007992cc0})
	GOROOT/src/testing/testing.go:1631 +0x1c4
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x33c
panic({0x4a573a0?, 0x4007992cc0?})
	GOROOT/src/runtime/panic.go:770 +0x124
github.com/cockroachdb/cockroach/pkg/sql/stats.randBounds(0x9683960, 0x40065def30, 0x8f)
	pkg/sql/stats/quantile_test.go:216 +0x554
github.com/cockroachdb/cockroach/pkg/sql/stats.randHist({0x6509e48, 0x9b66e60}, 0x9683960, 0x40065def30)
	pkg/sql/stats/quantile_test.go:71 +0x64
github.com/cockroachdb/cockroach/pkg/sql/stats.TestRandomQuantileRoundTrip.func1(0x400bbff040)
	pkg/sql/stats/quantile_test.go:45 +0x54
testing.tRunner(0x400bbff040, 0x400a734e70)
	GOROOT/src/testing/testing.go:1689 +0xec
created by testing.(*T).Run in goroutine 7079
	GOROOT/src/testing/testing.go:1742 +0x318
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomQuantileRoundTrip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/pgwire/pgwirebase/encoding.go
+++ b/pkg/sql/pgwire/pgwirebase/encoding.go
@@ -916,9 +916,13 @@ type PGNumeric struct {
 // for a timestamp. To create a timestamp from this value, it takes the microseconds
 // delta and adds it to PGEpochJDate.
 func pgBinaryToTime(i int64) time.Time {
+	// Postgres uses math.MaxInt64 microseconds as the "infinity" timestamp, see:
+	// https://github.com/postgres/postgres/blob/9380e5f129d2a160ecc2444f61bb7cb97fd51fbb/src/include/datatype/timestamp.h#L151
 	if i == math.MaxInt64 {
 		return pgdate.TimeInfinity
 	}
+	// Postgres uses math.MinInt64 microseconds as the "-infinity" timestamp, see:
+	// https://github.com/postgres/postgres/blob/9380e5f129d2a160ecc2444f61bb7cb97fd51fbb/src/include/datatype/timestamp.h#L150
 	if i == math.MinInt64 {
 		return pgdate.TimeNegativeInfinity
 	}

--- a/pkg/sql/sem/tree/datum.go
+++ b/pkg/sql/sem/tree/datum.go
@@ -85,12 +85,14 @@ var (
 	//
 	// Refer to the doc comments of the function "timeutil.Unix" for the process of
 	// deriving the arguments to construct a specific time.Time.
-	MaxSupportedTime = timeutil.Unix(9224318016000-1, 999999000) // 294276-12-31 23:59:59.999999
+	MaxSupportedTime    = timeutil.Unix(9224318016000-1, 999999000) // 294276-12-31 23:59:59.999999
+	MaxSupportedTimeSec = float64(MaxSupportedTime.Unix())
 	// MinSupportedTime is the minimum time we support parsing.
 	//
 	// Refer to the doc comments of the function "timeutil.Unix" for the process of
 	// deriving the arguments to construct a specific time.Time.
-	MinSupportedTime = timeutil.Unix(-210866803200, 0) // 4714-11-24 00:00:00+00 BC
+	MinSupportedTime    = timeutil.Unix(-210866803200, 0) // 4714-11-24 00:00:00+00 BC
+	MinSupportedTimeSec = float64(MinSupportedTime.Unix())
 )
 
 // CompareContext represents the dependencies used to evaluate comparisons
@@ -2597,7 +2599,7 @@ type DTimestamp struct {
 
 // MakeDTimestamp creates a DTimestamp with specified precision.
 func MakeDTimestamp(t time.Time, precision time.Duration) (_ *DTimestamp, err error) {
-	if t, err = checkTimeBounds(t, precision); err != nil {
+	if t, err = roundAndCheck(t, precision); err != nil {
 		return nil, err
 	}
 	return &DTimestamp{Time: t}, nil
@@ -2871,7 +2873,14 @@ type DTimestampTZ struct {
 	time.Time
 }
 
-func checkTimeBounds(t time.Time, precision time.Duration) (time.Time, error) {
+// roundAndCheck rounds the given time to the specified precision and checks
+// if the rounded time is within the supported bounds.
+//
+// Supported bounds:
+//   - [MinSupportedTime, MaxSupportedTime]
+//   - TimeInfinity
+//   - TimeNegativeInfinity
+func roundAndCheck(t time.Time, precision time.Duration) (time.Time, error) {
 	ret := t.Round(precision)
 	if ret.After(MaxSupportedTime) || ret.Before(MinSupportedTime) {
 		if t == pgdate.TimeInfinity || t == pgdate.TimeNegativeInfinity {
@@ -2885,7 +2894,7 @@ func checkTimeBounds(t time.Time, precision time.Duration) (time.Time, error) {
 
 // MakeDTimestampTZ creates a DTimestampTZ with specified precision.
 func MakeDTimestampTZ(t time.Time, precision time.Duration) (_ *DTimestampTZ, err error) {
-	if t, err = checkTimeBounds(t, precision); err != nil {
+	if t, err = roundAndCheck(t, precision); err != nil {
 		return nil, err
 	}
 	return &DTimestampTZ{Time: t}, nil
@@ -2928,7 +2937,7 @@ func ParseTimestampTZ(
 	if err != nil {
 		return time.Time{}, false, err
 	}
-	if t, err = checkTimeBounds(t, precision); err != nil {
+	if t, err = roundAndCheck(t, precision); err != nil {
 		return time.Time{}, false, err
 	}
 	return t, dependsOnContext, nil

--- a/pkg/sql/sem/tree/parse_string.go
+++ b/pkg/sql/sem/tree/parse_string.go
@@ -265,7 +265,7 @@ func ParseAndRequireStringHandler(
 		var ts time.Time
 		if ts, _, err = pgdate.ParseTimestampWithoutTimezone(now, dateStyle(ctx), s, dateParseHelper(ctx)); err == nil {
 			// Always normalize time to the current location.
-			if ts, err = checkTimeBounds(ts, TimeFamilyPrecisionToRoundDuration(t.Precision())); err == nil {
+			if ts, err = roundAndCheck(ts, TimeFamilyPrecisionToRoundDuration(t.Precision())); err == nil {
 				vh.TimestampTZ(ts)
 			}
 		}

--- a/pkg/sql/stats/quantile.go
+++ b/pkg/sql/stats/quantile.go
@@ -9,7 +9,6 @@ import (
 	"context"
 	"math"
 	"sort"
-	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/cat"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
@@ -376,20 +375,14 @@ func toQuantileValue(d tree.Datum) (float64, error) {
 		// converting back.
 		return float64(v.PGEpochDays()), nil
 	case *tree.DTimestamp:
-		if v.Equal(pgdate.TimeInfinity) {
-			return pgdate.TimeInfinitySec, nil
-		}
-		if v.Equal(pgdate.TimeNegativeInfinity) {
-			return pgdate.TimeNegativeInfinitySec, nil
+		if v.Before(tree.MinSupportedTime) || v.After(tree.MaxSupportedTime) {
+			return 0, tree.ErrFloatOutOfRange
 		}
 		return float64(v.Unix()) + float64(v.Nanosecond())*1e-9, nil
 	case *tree.DTimestampTZ:
 		// TIMESTAMPTZ doesn't store a timezone, so this is the same as TIMESTAMP.
-		if v.Equal(pgdate.TimeInfinity) {
-			return pgdate.TimeInfinitySec, nil
-		}
-		if v.Equal(pgdate.TimeNegativeInfinity) {
-			return pgdate.TimeNegativeInfinitySec, nil
+		if v.Before(tree.MinSupportedTime) || v.After(tree.MaxSupportedTime) {
+			return 0, tree.ErrFloatOutOfRange
 		}
 		return float64(v.Unix()) + float64(v.Nanosecond())*1e-9, nil
 	default:
@@ -461,15 +454,11 @@ func fromQuantileValue(colType *types.T, val float64) (tree.Datum, error) {
 		return tree.NewDDate(pgdate.MakeDateFromPGEpochClampFinite(int32(days))), nil
 	case types.TimestampFamily, types.TimestampTZFamily:
 		sec, frac := math.Modf(val)
-		var t time.Time
-		// Clamp to DTimestamp bounds.
-		if sec <= pgdate.TimeNegativeInfinitySec {
-			t = pgdate.TimeNegativeInfinity
-		} else if sec >= pgdate.TimeInfinitySec {
-			t = pgdate.TimeInfinity
-		} else {
-			t = timeutil.Unix(int64(sec), int64(frac*1e9))
+		// Return an error for all values outside the supported time range.
+		if sec < tree.MinSupportedTimeSec || sec > tree.MaxSupportedTimeSec {
+			return nil, tree.ErrFloatOutOfRange
 		}
+		t := timeutil.Unix(int64(sec), int64(frac*1e9))
 		roundTo := tree.TimeFamilyPrecisionToRoundDuration(colType.Precision())
 		if colType.Family() == types.TimestampFamily {
 			return tree.MakeDTimestamp(t, roundTo)

--- a/pkg/sql/stats/quantile_test.go
+++ b/pkg/sql/stats/quantile_test.go
@@ -193,17 +193,18 @@ func randBounds(colType *types.T, rng *rand.Rand, num int) tree.Datums {
 	case types.TimestampFamily, types.TimestampTZFamily:
 		roundTo := tree.TimeFamilyPrecisionToRoundDuration(colType.Precision())
 		var lo, hi int
-		if pgdate.TimeInfinitySec < math.MaxInt/2 {
-			lo = int(pgdate.TimeNegativeInfinitySec)
-			hi = int(pgdate.TimeInfinitySec)
+		if tree.MaxSupportedTimeSec < math.MaxInt/2 {
+			lo = int(tree.MinSupportedTimeSec)
+			hi = int(tree.MaxSupportedTimeSec)
 		} else {
-			// Make sure we won't overflow in randInts (i.e. make sure that
-			// hi - lo + 1 <= math.MaxInt which requires -2 for hi).
-			w := int(bits.UintSize) - 2
+			// Make sure we don't overflow in randInts on 32-bit systems.
+			// Specifically, make sure that hi - lo + 1 <= math.MaxInt, which requires subtracting 2 from hi.
+			w := bits.UintSize - 2
 			lo = -1 << w
 			hi = (1 << w) - 2
 		}
 		secs := randInts(num, lo, hi)
+
 		for i := range datums {
 			t := timeutil.Unix(int64(secs[i]), 0)
 			var err error
@@ -794,12 +795,12 @@ func TestQuantileValueRoundTrip(t *testing.T) {
 		{
 			typ: types.Timestamp,
 			dat: &tree.DTimestamp{Time: pgdate.TimeInfinity},
-			val: pgdate.TimeInfinitySec,
+			err: true,
 		},
 		{
 			typ: types.Timestamp,
 			dat: &tree.DTimestamp{Time: pgdate.TimeNegativeInfinity},
-			val: pgdate.TimeNegativeInfinitySec,
+			err: true,
 		},
 		{
 			typ: types.TimestampTZ,
@@ -809,12 +810,12 @@ func TestQuantileValueRoundTrip(t *testing.T) {
 		{
 			typ: types.TimestampTZ,
 			dat: &tree.DTimestampTZ{Time: pgdate.TimeInfinity},
-			val: pgdate.TimeInfinitySec,
+			err: true,
 		},
 		{
 			typ: types.TimestampTZ,
 			dat: &tree.DTimestampTZ{Time: pgdate.TimeNegativeInfinity},
-			val: pgdate.TimeNegativeInfinitySec,
+			err: true,
 		},
 	}
 	ctx := context.Background()
@@ -1049,50 +1050,43 @@ func TestQuantileValueRoundTripOverflow(t *testing.T) {
 		{
 			typ: types.Timestamp,
 			val: float64(pgdate.TimeNegativeInfinity.Unix()),
-			dat: &tree.DTimestamp{Time: pgdate.TimeNegativeInfinity},
-			res: pgdate.TimeNegativeInfinitySec,
+			err: true,
 		},
 		{
 			typ: types.Timestamp,
 			val: float64(pgdate.TimeInfinity.Unix()),
-			dat: &tree.DTimestamp{Time: pgdate.TimeInfinity},
-			res: pgdate.TimeInfinitySec,
+			err: true,
 		},
 		{
 			typ: types.Timestamp,
 			val: -math.MaxFloat64,
-			dat: &tree.DTimestamp{Time: pgdate.TimeNegativeInfinity},
-			res: pgdate.TimeNegativeInfinitySec,
+			err: true,
 		},
 		{
 			typ: types.Timestamp,
 			val: math.MaxFloat64,
-			dat: &tree.DTimestamp{Time: pgdate.TimeInfinity},
-			res: pgdate.TimeInfinitySec,
+			err: true,
 		},
+		// TimestampTZ cases.
 		{
 			typ: types.TimestampTZ,
 			val: float64(pgdate.TimeNegativeInfinity.Unix()),
-			dat: &tree.DTimestampTZ{Time: pgdate.TimeNegativeInfinity},
-			res: pgdate.TimeNegativeInfinitySec,
+			err: true,
 		},
 		{
 			typ: types.TimestampTZ,
 			val: float64(pgdate.TimeInfinity.Unix()),
-			dat: &tree.DTimestampTZ{Time: pgdate.TimeInfinity},
-			res: pgdate.TimeInfinitySec,
+			err: true,
 		},
 		{
 			typ: types.TimestampTZ,
 			val: -math.MaxFloat64,
-			dat: &tree.DTimestampTZ{Time: pgdate.TimeNegativeInfinity},
-			res: pgdate.TimeNegativeInfinitySec,
+			err: true,
 		},
 		{
 			typ: types.TimestampTZ,
 			val: math.MaxFloat64,
-			dat: &tree.DTimestampTZ{Time: pgdate.TimeInfinity},
-			res: pgdate.TimeInfinitySec,
+			err: true,
 		},
 	}
 	ctx := context.Background()

--- a/pkg/util/timeutil/pgdate/parsing.go
+++ b/pkg/util/timeutil/pgdate/parsing.go
@@ -73,13 +73,9 @@ var (
 	// resulted in '23:59:59.999999'.) This behavior may change in the future, see issue #129148
 	// for more details.
 	//
-	// Postgres uses math.MaxInt64 microseconds as the infinity value.
-	// See: https://github.com/postgres/postgres/blob/9380e5f129d2a160ecc2444f61bb7cb97fd51fbb/src/include/datatype/timestamp.h#L157
-	//
 	// Refer to the doc comments of the function "timeutil.Unix" for the process of
 	// deriving the arguments to construct a specific time.Time.
-	TimeInfinity    = timeutil.Unix(9224318102399 /* sec */, 999999000 /* nsec */)
-	TimeInfinitySec = float64(TimeInfinity.Unix())
+	TimeInfinity = timeutil.Unix(9224318102399 /* sec */, 999999000 /* nsec */)
 	// TimeNegativeInfinity represents the "lowest" possible time. Its value is
 	// "-4714-11-23 00:00:00 +0000 UTC", which is 24 hours before "MinSupportedTime"
 	// ("-4714-11-24 00:00:00 +0000 UTC").
@@ -92,13 +88,9 @@ var (
 	// resulted in '00:00:00'.) This behavior may change in the future, see issue #129148
 	// for more details.
 	//
-	// Postgres uses math.MinInt64 microseconds as the -infinity value.
-	// See: https://github.com/postgres/postgres/blob/9380e5f129d2a160ecc2444f61bb7cb97fd51fbb/src/include/datatype/timestamp.h#L156
-	//
 	// Refer to the doc comments of the function "timeutil.Unix" for the process of
 	// deriving the arguments to construct a specific time.Time.
-	TimeNegativeInfinity    = timeutil.Unix(-210898425600 /* sec */, 0 /* nsec */)
-	TimeNegativeInfinitySec = float64(TimeNegativeInfinity.Unix())
+	TimeNegativeInfinity = timeutil.Unix(-210898425600 /* sec */, 0 /* nsec */)
 )
 
 type ParseHelper struct {

